### ISSUE TITLE
support parameter in test

### ISSUE THREAD

**Issue by jencce** [teemtee/tmt#1124]
For example, we have a test named kernelinstall, which is used to install newer kernel
before testing.
Than we have a plan, for example tier1, which is consist of those tests:

- name : kernelinstall
   test: ...
- name: test1
  test: ...
...

If we have to hard code the kernel name version release in the plan or the test, it would
be very annoying because we need to modify/commit/push for every run, that means
hundreds of commits for each release.

Any suggestions are appreciated. Thanks!

**Comment by lukaszachy** [comment]
You can set environment variables via cmdline, will that help? https://tmt.readthedocs.io/en/stable/spec/plans.html#environment

**Comment by jencce** [comment]
Ya, environment variables can do this. Thanks!

**Comment by jencce** [comment]
During migration, I found a scenario is not properly supported by environment variables for parameterized test.

In a same plan, I run one same testsuite multiple times, with different parameters, to test different filesystems types.  Environment variables can't do this.

This scenario is important for our usage. Please help. Reopening.

Thanks!

**Comment by emmawu** [comment]
Is that possible to support the environemnt variables setting on plan, instead of test itself? 

For example, we have a kexec-boot test and it allows uers to pass in whatever kernel options during execution. So the test script itself doesn't know what kernel option it will be running on. It is the "Test Plan" to decide how many times we'd like to run kexec-boot repeatedly and what kernel options we'd like to pass. In this way, the test parameters (scenarios) are maintained in test plan, not hard coded in test scrip itself.

**Comment by psss** [comment]
Yes, it's possible to set environment variables in both test and plan:

* https://tmt.readthedocs.io/en/latest/spec/tests.html#environment
* https://tmt.readthedocs.io/en/latest/spec/plans.html#environment

Also note that for a single test script there can multiple tmt tests:

```yaml
description: Check download options
test: ./test.sh
require: curl

/fast:
    description: Check download options (quick smoke test)
    tier: 1
    duration: 1m
    environment:
        MODE: fast

/full:
    description: Check download options (full test set)
    tier: 2
    duration: 10m
    environment:
        MODE: full
```

From the plan you can then filter appropriate set of tests as desired.

**Comment by psss** [comment]
Another approach could be a single tmt test and multiple tmt plans. Here's a simple working example:

```yaml
/test:
    test: echo $VARIABLE

/plan:
    discover:
        how: fmf
    provision:
        how: local
    execute:
        how: tmt

    /one:
        summary: The first plan
        environment:
            VARIABLE: 1

    /two:
        summary: The second plan
        environment:
            VARIABLE: 2
```

Thanks to `fmf` you can apply virtual hierarchy to inherit common plan stuff and only set the environment.

**Comment by jencce** [comment]
Thank you very much Petr! Very helpful!

For setting environment variables in test, you mean set in testcase fmf file.
For setting in plan, your latest example works but it ends up with 2 plans which means testing in 2 jobs. This is kind of working for our requirements.

But for the best of our usage, first we don't want to writing hard coded parameter values in testcase fmf file, because it's hard to maintain thousands of testcases; second, it's the best we can repeat the same testcase with different parameters in 1 plan, like:
```
discover:
    - name: some-tests
      how: fmf
      url: https://xxx/tests.git
      ref: master
      test:
          - /general/reading-test
            environment:
                FSTYP: xfs
          - /general/reading-test
            environment:
                FSTYP: ext4
provision:
    how: connect
execute:
    how: tmt
    framework: shell
```

It now fails with parsing error:
```
mapping values are not allowed in this context
```

Could it be possible to support this in the future?

**Comment by lukaszachy** [comment]
@jencce That error message is from yaml parser - the document is not a valid yaml document (mixing str and dict as a value for list). There is nothing for tmt to do.


The problem is in 
```yaml
      test:
          - /general/reading-test
            environment:
                FSTYP: xfs
```
Correct document is 
```
      test:
          - name: /general/reading-test
            environment:
                FSTYP: xfs
          - name: /general/reading-test
            environment:
                FSTYP: ext4
```

**Comment by jencce** [comment]
Thanks @lukaszachy !  I can run the plan now but get error like this:

# STDERR:
---v---v---v---v---v---
Traceback (most recent call last):
  File "/usr/bin/tmt", line 8, in <module>
    tmt.cli.main()
  File "/usr/lib/python3.6/site-packages/click/core.py", line 721, in __call__
    return self.main(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 696, in main
    rv = self.invoke(ctx)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 1065, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/lib/python3.6/site-packages/click/core.py", line 1092, in invoke
    return _process_result(rv)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 1030, in _process_result
    **ctx.params)
  File "/usr/lib/python3.6/site-packages/click/core.py", line 534, in invoke
    return callback(*args, **kwargs)
  File "/usr/lib/python3.6/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/usr/lib/python3.6/site-packages/tmt/cli.py", line 384, in finito
    click_context.obj.run.go()
  File "/usr/lib/python3.6/site-packages/tmt/base.py", line 1822, in go
    plan.go()
  File "/usr/lib/python3.6/site-packages/tmt/base.py", line 1040, in go
    step.go()
  File "/usr/lib/python3.6/site-packages/tmt/steps/discover/__init__.py", line 154, in go
    phase.go()
  File "/usr/lib/python3.6/site-packages/tmt/steps/discover/fmf.py", line 455, in go
    excludes=excludes)
  File "/usr/lib/python3.6/site-packages/tmt/base.py", line 1410, in tests
    in name_filter(self.tree.prune(keys=keys, names=[name]))]
  File "/usr/lib/python3.6/site-packages/tmt/base.py", line 1409, in <listcomp>
    Test(test) for test
  File "/usr/lib/python3.6/site-packages/fmf/base.py", line 563, in prune
    [re.search(name, node.name) for name in names]):
  File "/usr/lib/python3.6/site-packages/fmf/base.py", line 563, in <listcomp>
    [re.search(name, node.name) for name in names]):
  File "/usr/lib64/python3.6/re.py", line 182, in search
    return _compile(pattern, flags).search(string)
  File "/usr/lib64/python3.6/re.py", line 289, in _compile
    p, loc = _cache[type(pattern), pattern, flags]
TypeError: unhashable type: 'dict'

---^---^---^---^---^---

**Comment by lukaszachy** [comment]
Ouch. I see it now. You are using `discover --how fmf` but that wants only NAME of the test (https://tmt.readthedocs.io/en/stable/spec/plans.html#fmf). I got confused because `discover --how shell`  allows most of test attributes to be defined there (I missed the 'how: fmf')

In near future `shell` will be able to fetch tests from url (#1624) however using that you would loose other fmf metadata (mainly require/recommend).

So basically what you are requesting is for `discover fmf` to be able to override discovered metadata. 
@psss I think that might be handy and should be too difficult to implement. The only issue is how to make writing the plan user friendly.

**Comment by kkaarreell** [comment]
I believe I have filed similar RFE https://github.com/teemtee/tmt/issues/1793

**Comment by psss** [comment]
@jencce, sorry for the late response, is your issue still relevant? Here's what is currently possible using the `adjust-tests` key:

```yaml
/plan:
    provision:
        how: local
    discover:
      - how: fmf
        adjust-tests:
            environment:
                VAR: 1
      - how: fmf
        adjust-tests:
            environment:
                VAR: 2
      - how: fmf
        adjust-tests:
            environment:
                VAR: 3
    execute:
        how: tmt

/test:
    test: echo I got $VAR
```

See also the [Adjust Tests](https://tmt.readthedocs.io/en/stable/plugins/discover.html#adjust-tests) section for more information. Does that help with your use case? Anyway I'd suggest to continue with #1793 to track support for including test specific environment variables directly under the `test` key.

**Comment by jencce** [comment]
Yes, "Adjust Tests" looks helpful. Agree with you that env variables directly under test key would be good. Thanks for driving this!

**Comment by psss** [comment]
Ok, let's track the per-test-variables functionality in #1793.
